static-analysis/teammates-macker.xml

```diff
@@ -26,20 +181,57 @@
             </deny>
         </access-rule>
     </ruleset>
-    <ruleset name="Only UI tests can access page object classes">
+    <ruleset name="TestDriverComponent: Only UI tests can access page object classes">
         <access-rule>
             <deny>
                 <from class="teammates.test.**" />
-                <to class="teammates.test.pageobjects.*" />
+                <to class="teammates.test.pageobjects.**" />
                 <allow>
                     <!-- allow dependency within the package -->
-                    <from class="teammates.test.pageobjects.*" />
+                    <from class="teammates.test.pageobjects.**" />
                 </allow>
                 <allow>
                     <!-- allow dependency from UI tests -->
-                    <from class="teammates.test.cases.browsertests.*" />
+                    <from class="teammates.test.cases.browsertests.**" />
                 </allow>
             </deny>
         </access-rule>
     </ruleset>
+    <ruleset name="TestDriverComponent: Test driver can only be accessed from test package">
+        <access-rule>
+            <deny>
+                <from class="teammates.**" />
+                <to class="teammates.test.driver.**" />
+                <allow>
+                    <from class="teammates.test.**" />
+                </allow>
+                <allow>
+                    <from class="teammates.client.**" />
+                </allow>
+            </deny>
+        </access-rule>
+    </ruleset>
+    <ruleset name="TestDriverComponent: util and browsertest cannot access GaeSimulation">
+        <access-rule>
+            <deny>
+                <from>
+                    <include class="teammates.test.cases.util.**" />
+                    <include class="teammates.test.cases.browsertests.**" />
+                </from>
+                <to class="teammates.test.driver.GaeSimulation" />
+            </deny>
+        </access-rule>
+    </ruleset>
+<!--     This rule need to be enforced gradually. -->
+<!--     <ruleset name="TestDriverComponent: action test cases should only interact with back-end through BackDoor API">
```

Comment:
OK for this one, let's get it straight.
- Any class that uses `GaeSimulation` can access `BackDoorLogic` directly, so you might actually be able to enforce together with the previous rule.
- BackDoor API is meant for browser tests, essentially to hide the code behind `BackDoorLogic`. All other tests should not have the need for it, with the sole exception of `BackDoorTest` for obvious reason.
- Any other test does not have anything to do with data and thus should not be able to access `BackDoorLogic`, `BackDoor`, `GaeSimulation`.